### ISSUE TITLE
Constraints on optimization

### ISSUE THREAD

**Issue by lars-t-hansen** [tc39/ecmascript_sharedmem#24]
An example that came up on a mailing list:

``` js
// untrusted code
1  var tmp = <load from shared memory>;  // races with a write

// trusted code
2  if (SecurityCheck(tmp))
3    doSomething(tmp)
```

Suppose the load on line 1 races with a write and that the call in line 2 needs to spill the register containing tmp.  Under some assumptions of analysis (eg inlining) the optimizer might discover that SecurityCheck does not affect the shared memory, so the compiler might generate code to to reload tmp on line 3, instead of saving it across the call on line 2.  But that value may now have changed because the racing write changed it after it was read the first time.

It's probably the case that ES2015 contains wording that prohibits this, but it's also probably the case that the JITs have optimizers that performs that optimization since the reload is not observable in single-threaded code.

It may be the case that such optimizations need to be outlawed when the load is from shared memory.  (This as an example of a broader class of issues.)  If the memory is unshared then there should be no impact.

**Comment by lars-t-hansen** [comment]
The re-loading is arguably outlawed by the current spec since a read from a racy location may return a non-identical value (it is not UB), ergo the optimization does not preserve behavior.  Even so I have added a note to the section on GetValueFromBuffer highlighting the issue for existing optimizers.

**Comment by lars-t-hansen** [comment]
Fixed in [master 00bf307].
